[nova] Add database nanny to Nova

seb-kro opened the pull request:
We include the Nova database maintenance tasks from the`openstack/nannies` chart directly into the Nova chart via new cronjobs.

Both the previous nanny and the new nanny run the commands `archive_deleted_rows` and `purge` from `nova-manage db`. Unlike the previous nanny, the new nanny passes the complete daytime as the `before` parameter to these commands instead of the rounded date. That way the cleanup is performed in smaller batches if the nanny runs more than once a day.

The previous nanny additionally executed two Python scripts from [https://github.com/sapcc/openstack-nannies/tree/master/scripts](https://github.com/sapcc/openstack-nannies/tree/master/scripts) for further cleanup.

The previous script `nova-queens-instance-mapping.py` was only executed in dry-run mode without performing any changes in all regions other than qa-de-1. We therefore skip the corresponding cell mapping in the `instance_mappings` for now, as well as the cleanup of the
`build_requests` table. We are still planning to implement the instance mapping as a next step due to a recent issue with a missing cell ID in `instance_mappings`.

The previous script `nova-consistency.py` performed various other cleanup tasks.

The soft-deletion of leftover BDMs for already deleted volumes is no longer required since we are now on the Bobcat release which implements corresponding cleanup upon instance reboot via `_delete_dangling_bdms` in `nova/compute/manager.py` (https://docs.openstack.org/releasenotes/nova/2023.2.html, https://specs.openstack.org/openstack/nova-specs/specs/2023.2/implemented/cleanup-dangling-volume-attachments.html)

Function `purge_instance_faults` is replaced by the newly implemented `nova-manage db soft_delete_excessive_instance_faults` command: https://github.com/sapcc/nova/pull/552

Custom purging of tables `block_device_mapping`, `reservations` and `instance_id_mappings` is no longer required because testing shows that the commands `archive_deleted_rows` and `purge` do now consider these tables as well.

joker-at-work commented on openstack/nova/templates/nanny-db-cronjob.yaml:
```diff
@@ -0,0 +1,135 @@
+                    path: nova.conf
+                  - key:  nova-api.conf
+                    path: nova-api.conf
+                  - key:  api-paste.ini
+                    path: api-paste.ini
+                  - key:  policy.yaml
+                    path: policy.yaml
```
Unless we run an API, we don't need these.

joker-at-work commented on openstack/nova/templates/nanny-db-cronjob.yaml:
```diff
@@ -0,0 +1,135 @@
+                    path: policy.yaml
+                  - key:  logging.ini
+                    path: logging.ini
+                  - key:  rootwrap.conf
+                    path: rootwrap.conf
+                  - key:  api-metadata.filters
+                    path: rootwrap.d/api-metadata.filters
```
we do not need those either.

seb-kro replied:
Removed.

joker-at-work commented on openstack/nova/templates/nanny-db-cronjob.yaml:
```diff
@@ -0,0 +1,135 @@
+                    path: rootwrap.d/api-metadata.filters
+                  - key:  release
+                    path: release
+                  {{- if .Values.audit.enabled }}
+                  - key:  nova_audit_map.yaml
+                    path: nova_audit_map.yaml
+                  {{- end }}
+                  {{- if .Values.watcher.enabled }}
+                  - key:  watcher.yaml
+                    path: watcher.yaml
+                  {{- end }}
```
These are only for APIs

seb-kro replied:
Removed.

joker-at-work reviewed (commented):
I'd like to split the nanny commands into different Cronjobs, so we can have different schedules for them and one thing breaking doesn't break all of them.

We could then also get rid of the shell variables and do the templating in helm.

seb-kro commented on openstack/nova/templates/bin/_nova-nanny-db.tpl:
```diff
@@ -0,0 +1,49 @@
+        echo "ERROR: invalid boolean string: $str"
+        exit 1
+    fi
+}
```
Adjusted.

seb-kro commented on openstack/nova/values.yaml:
```diff
@@ -1013,3 +1013,28 @@ vpa:
+    schedule: "*/15 * * * *"
+    # For each instance, soft-delete the instance_faults db rows that exceed the
+    # count of max_faults. Use batch size max_rows to limit query size.
+    soft_delete_instance_faults:
```
Adjusted.

seb-kro commented on openstack/nova/values.yaml:
```diff
@@ -1013,3 +1013,28 @@ vpa:
+      max_rows: 25
+    # Archive db rows soft-deleted before more than older_than days to shadow
+    # tables. Use batch size max_rows limit transaction size.
+    archive:
```
Adjusted.

joker-at-work reviewed (changes requested):
* We're missing linkerd integration.
* Please add newlines at the end of your new files.

seb-kro commented on openstack/nova/values.yaml:
```diff
@@ -1013,3 +1013,29 @@ vpa:
+  # Archive db rows soft-deleted before more than older_than days to shadow
+  # tables. Use batch size max_rows to limit transaction size.
+  db_archive_deleted_rows:
+    enabled: false
```
Thought that it was preferred to enable at region level / in the secrets. Adjusted.

seb-kro commented on openstack/nova/templates/bin/_nanny-db-archive-deleted-rows.tpl:
```diff
@@ -0,0 +1,12 @@
+#!/usr/bin/env bash
+set -euo pipefail
+
+archive_before=$(date --date "$(date -Iseconds -u) - {{ .Values.nanny.db_archive_deleted_rows.older_than }} days" \
```
Adjusted.

seb-kro commented on openstack/nova/templates/nanny-db-archive-deleted-rows-cronjob.yaml:
```diff
@@ -0,0 +1,93 @@
+          restartPolicy: OnFailure
+          {{- include "utils.proxysql.job_pod_settings" . | indent 10 }}
+          volumes:
+          - name: container-init
+            configMap:
+              name: nova-bin
+              defaultMode: 0755
```
Adjusted in the `volumes` configuration.

seb-kro commented:
Added the newlines. However, I believe linkerd is already configured? The `job_metadata` sets linkerd annotations by including `utils.linkerd.pod_and_service_annotation`. And `utils.script.job_finished_hook` terminates the linkerd sidecar. Or maybe I am overlooking sth?

As discussed with @joker-at-work, I also removed proxysql because this was causing query timeouts in `qa-de-1` for the `purge` nanny. And it is okay for the nanny to lose database connection and fail, since the next run will simply resume database cleanup.

For the `archive` nanny, I further increased `max_rows` to 250, which is still well below the default value of 1000. Due to a bug, on `qa-de-1` the previous nanny was no longer running for some weeks so there was a lot of data to archive. And the `archive` command is not very efficient and reiterates over many already archived tables between batches. So increasing batch size helped and should prevent long-running jobs and job failures in similar situations.

seb-kro commented on openstack/nova/templates/bin/_nanny-db-purge.tpl:
```diff
@@ -0,0 +1,10 @@
+#!/usr/bin/env bash
+set -euo pipefail
+
+purge_before=$(date --date "$(date -Iseconds -u) - {{ .Values.nanny.db_purge.older_than }} days" \
```
Adjusted.

seb-kro commented on openstack/nova/templates/nanny-db-archive-deleted-rows-cronjob.yaml:
```diff
@@ -0,0 +1,93 @@
+                    path: nova-{{ .Values.cell2.name }}.conf
+                  {{- end }}
+                  - key: keystoneauth-secrets.conf
+                    path: nova.conf.d/keystoneauth-secrets.conf
```
Tested and removed.

seb-kro commented on openstack/nova/templates/nanny-db-archive-deleted-rows-cronjob.yaml:
```diff
@@ -0,0 +1,93 @@
+                  {{- if .Values.audit.enabled }}
+                  - key: audit-middleware.conf
+                    path: nova.conf.d/audit-middleware.conf
+                  {{- end }}
```
Removed.

joker-at-work commented on openstack/nova/templates/nanny-db-archive-deleted-rows-cronjob.yaml:
```diff
@@ -0,0 +1,76 @@
+                    path: nova.conf
+                  - key:  logging.ini
+                    path: logging.ini
+                  - key:  release
+                    path: release
```
minor: Do you know if `release` is used? We recently had that same question with `nova-compute` and it seemed unused there.

seb-kro replied:
Will check when implementing the template and adding the planned instance mapping nanny.

joker-at-work commented on openstack/nova/templates/nanny-db-archive-deleted-rows-cronjob.yaml:
```diff
@@ -0,0 +1,76 @@
+            {{- include "utils.trust_bundle.volume_mount" . | indent 12 }}
+            resources:
+              requests:
+                memory: "500Mi"
+                cpu: "100m"
```
minor: these being hard-coded means we cannot increase them in busy regions. they should be read from values.

seb-kro replied:
Will adjust when implementing the template and adding the planned instance mapping nanny.

joker-at-work reviewed (approved):
I'm fine with this, but we should probably follow-up.

We could think about defining a Job template for the nanny jobs, as at least the current ones all use the same Job with just the name and some settings changed. We could pass both into an `include` that renders the Job.
